FLEDGE.md

```diff
@@ -109,9 +119,15 @@ The browser will remain in an interest group for only a limited amount of time.
 
 The `userBiddingSignals` is for storage of additional metadata that the owner can use during on-device bidding, and the `trustedBiddingSignals*` attributes provide another mechanism for making real-time data available for use at bidding time.
 
-The `dailyUpdateUrl` provides a mechanism for the group's owner to periodically update the attributes of the interest group: any new values returned in this way overwrite the values previously stored (except that the `name` and `owner` cannot be changed).  However, the browser will only allow daily updates when a sufficiently large number of people have the same `dailyUpdateUrl` , e.g. at least 100 browsers with the same update url. This will not include any metadata, so data such as the interest group `name` should be included within the url, so long as url exceeds the minimum count threshold.  (Without this sort of limit, a single-person interest group could be used to observe that person's coarse-grained IP-Geo location over time.)
+The `dailyUpdateUrl` provides a mechanism for the group's owner to periodically update the attributes of the interest group. An `updatedInterestGroup` object obtained in this way will be used to update the current `interestGroup` in the following way:
+- `name` and `owner` fields cannot be changed
+- `adComponents` will be "shallowly merged": each key and value in `updatedInterestGroup.adComponents` will overwrite the previous value in `interestGroup.adComponents` (or set it, if it was not present previously). This way the owner has the flexibility to update some ads / products without overwriting others, for example to advertise a newly released product. (See issue #160.)
```

Comment:
Two concerns about shallow merges here:
1) Having more separately updateable fields makes the potential for version skew within an interest group much more likely. Even though interest groups are supposed to be updated daily, they will not be updated when the browser is not running, there is no network access, the server is unreachable, FLEDGE is disabled, etc. With a shallow merge there is are no guarantees about the resulting client state, so bidding scripts need to be able to handle any possible combination of previous updates.
2) There is no mechanism specified for dropping one of these custom key groups. Over time these will likely accumulate (as long as the interest group still exists), requiring more client-side resources.

Reply:
Hi @brusshamilton,

1. I'd say it's the buyer's responsibility to use this mechanism so that the bidding script works well. (I.e. the buyer should not assume anything about the success of previous daily updates.)

2. For dropping: my natural inclination would be to overwrite a key with an empty list (but we could add a dedicated mechanism for removing a key as well). As to accumulation - I think here too it's best to put the responsibility on the buyer. It seems some kind of a limit on the size of an interest group is needed anyway. If a daily update would exceed that limit, the update fails. A hard limit on the number of keys / total number of ad components could be added too, if useful.

Note that our core use case doesn't have these problems: the daily update would simply overwrite the list of newly released products.

Reply:
1. I agree that it is the interest group owner's responsibility to make sure their bidding script works, but I think we should consider how difficult we make the API to use correctly. Previously the number of data versions a bidding script needed to test and support was linear with the number of changes. With separately updateable fields the number of versions is combinatorial with the number of changes. Are there for benefits of having separately updateable fields other than the obvious bandwidth savings?
2. Yes, the browser should limit interest group size and, yes, the contents of the interest group are the interest group owner's responsibility, but I'm concerned it will be difficult for interest group owners to detect and correct errors. Errors from interest group updates are not exposed to the owner so there is no way for the owner to know if an update failed on a set of users. Without partial updates then any update that would fail would also fail on join (which is easier to test and debug). With partial updates such failures are dependent on client state, which we likely can not provide to interest group owners while meeting the needed privacy constraints.

Your current use case may not have those problems, but your use case may change. Others may not share your use case. In order for FLEDGE to be useful as a web platform features we need to consider all users.

Reply:
> Are there for benefits of having separately updateable fields other than the obvious bandwidth savings?

Yes, sorry if this wasn't clear from the beginning - the purpose of this pull req is not to optimize bandwidth, but to enable new functionalities:
1. Update "ads" without updating "products".
    - The idea behind [Product-level TD](https://github.com/WICG/turtledove/blob/main/PRODUCT_LEVEL.md) ("PLTD") was that a set of products can be specified for each user at the moment of `joinAdInterestGroup` call. This *set* of products can potentially be pretty unique, while individual products should still pass k-anonymity thresholds to prevent microtargeting. This concept is crucial to achieve satisfying performance for cases where the advertiser is offering multiple products (mostly ecommerce). (PLTD has been accepted and incorporated into the FLEDGE.) However, the technical details of the current FLEGE specification prevent advertisers from using this mechanizm fully and efficiently. The problem is that the `ads` field mixes 2 things: ads (that in an ecommerce case are typically *templates*) and products (that fill these ad templates). Now, if we update the `ads` field, in order to update *ad templates*, we will lose the original product set specified while joining the IG, as the daily update is subject to a popularity threshold (and so does not have access to any user identity). In order to follow the idea originally described in PLTD we would need to be able to keep the list of original products specified in `joinAdInterestGroup` and daily-update only the ads (*templates*). 
    - (This is also described in issue #160)
2. Add fresh "most popular" products without overwriting other products in the interest group.
    - This way, once a new iPhone comes out, we may update existing Interest Groups, and advertise it along with phone recommendations already present in each Interest Group.
3. In use case 2, we may also like to extend the trusted bidding signal keys so that they contain ids of the new products.
    - This way, during bidding, we may check the availability of each product.

It seems that each such use case is easily solved by adding certain fields in the interest group's toplevel namespace. (Note that the daily update, as proposed in FLEDGE, performs a shallow merge within this namespace.)

So we could split the current `ads` field into `ads` and `products`; then we could add an additional `freshProducts` field, and also `freshProductTrustedBiddingSignalsKeys`, ...

It seems though that FLEDGE doesn't need to know about all these fields, rather it could provide a flexible mechanism for updating certain fields "partially" in a way controlled by the buyer. This led me to the idea of a "shallow merge" at a field level (as opposed to the top level), as described in this pull req.

I see your point about complicating the API. I guess there are a couple options:
- a) From our perspective, use case 1 is the most important. We could start with only solving this use case, by splitting the field into two: `ads` and `products`. (I'm quite sure other use cases will follow, but maybe it’s best to add support for them later, and only solve the key use case now in order to keep the API simple?)
- b) Perhaps we could seek other ways to support flexible updates, without complicating the API.
    - One approach would be to allow the **optional** specification of a `dailyUpdateCustomLogic`, where the buyer could perform a custom merge based on the current IG, and the data from the daily update response. (So if `dailyUpdateCustomLogic` is not provided, the daily update works by overwriting respective fields, as originally described in FLEDGE.) 

Please let me know if this makes sense, and what do you think would be the best approach. Should we go with a) for now?

Reply:
With the current FLEDGE specification, it seems like the set of "products" could be set in the "userBiddingSignals" provided when the user joins the interest group. This field can only be updated by joinAdInterest group so it seems like it would be a good fit for holding the per-user "products" list.

That alone may not be sufficient for your use case, since it looks like you need to store some per-interest group data separate from the "ads" field (since entries in that list are required to correspond to a creative or template). I think it may make sense to add an additional top level field to the interest group, but maybe we could call it "config" instead of "products" since it is likely it could be used for other use cases. @michaelkleber What do you think?

Reply:
> With the current FLEDGE specification, it seems like the set of "products" could be set in the "userBiddingSignals" provided when the user joins the interest group. 

Note that the browser will need clarity about "what products are stored in an interest group", so the products will have to be stored in a structured, browser-interpretable field (as opposed to the userBiddingSignals, which are custom, owner-specified, not interpretable by the browser).

From that perspective, products are very similar to ads. For example, the browser needs to be able to remove rare ads from the interest group in order to enforce microtargeting prevention (see last paragraph in [section 3.3](https://github.com/WICG/turtledove/blob/main/FLEDGE.md#33-metadata-with-the-ad-bid)), and this mechanism will have to be applied to products as well, see [section 3.4](https://github.com/WICG/turtledove/blob/main/FLEDGE.md#34-ads-composed-of-multiple-pieces):

> (...) This is useful because the browser's microtargeting threshold can be applied to each individual component of the ad without compromising on tracking protections.

Reply:
Right, I think I agree with @jonasz here.  What he's calling "products" are intended to be pieces of an ad, which appear inside other ads that act as containers.  That means that the URLs associated with products are intended to go through all the same k-anonymity checking as anything else that we plan to show on the screen, for example.  So this does deserve some structure that is more like `ads` rather than `userBiddingSignals`.

Reply:
@brusshamilton shall we go with `ads` and `adsComponents` in that case? If that sounds good to you, I'll update the pull req.

Reply:
I still have concerns about the 'shallow merge'. I'm concerned that this prevents interest group owners from cleaning up after themselves when they make breaking changes, and also doesn't provide a standardized way for browsers to identify and remove parts of the dictionary that are no longer used.

This also has the potential to compound the issue in #188, where partial updates of interest groups may allow multiple sites to inject information into an ad auction.

Reply:
Hi Russ - as proposed in https://github.com/WICG/turtledove/pull/185#discussion_r651082826, option a - let's just split `ads` into `ads` and `adComponents`. This solves use case 1. This option assumes no shallow merge at all - sorry if this wasn't clear!

Does that sound good to you?

Reply:
If there's no shallow merge then there shouldn't be a problem. I think I was confused because the latest version still says it is a shallow merge.

Reply:
I see - I didn't want to update the spec without first agreeing on what should be done. Please take a look, I updated the spec now. Let me know if you'd prefer the three commits to be squashed into one.

Reply:
The spec looks fine to me now. I think it would be better to squash it into a single commit.
@michaelkleber Does this seem fine to you?

Reply:
Yes, LGTM

Reply:
LGTM